pilot/pkg/security/authz/policy/helper.go

```diff
@@ -229,3 +229,32 @@ func Verify(got *envoy_rbac.RBAC, want map[string][]string) error {
 
 	return err
 }
+
+// checkSourcePrincipalCounts checks if |actual| and |want| have the same number of source.principal.
+// For example, both |actual| and |want| can only have one key "source.principal",
+// but |actual| can have "foo" and "bar", and |want| can only has "foo". This is possible since
+// in v1beta1, principals can be a list.
+func checkPrincipalCounts(ruleKey string, actual string, want []string) error {
```

Comment:
Why do we need this check?

1.  This function hard codes the `source.principal` which seems assuming the `want` is always `principal` but that is not the purpose of these unit tests here.

2. This change is also breaking the convention in the unit test, the unit test is using the `want` as a list of policy placeholder that we want to find in the `actual`.

Reply:
We need this check because the current Verify() function is not exhaustive. 
https://github.com/istio/istio/blob/6840ad6cd8d763753ed195e7baba949baa33fb58/pilot/pkg/security/authz/policy/helper.go#L222

Here we check if `actualStr` contains `value`, we didn't check the other way around. Consider this case:

`values := ["source.principal: foo"]`, then
`actualStr: source.principal: foo, source.principal: bar`, `value: source.principal: foo`. The function will succeed even though it should not.

Reply:
To give you a more detailed example:

```
{
			name: "one policy",
			policies: []model.Config{
				*policy.SimpleAuthzPolicy("default", "foo",
					[]string{policy.CustomPrincipal("cluster.local", "foo", "bar")}),
			},
			trustDomain:        "td",
			trustDomainAliases: []string{"cluster.local"},
			wantRules: map[string][]string{
				"ns[foo]-policy[default]-rule[0]": {
					policy.CustomPrincipal("cluster.local", "foo", "bar"),
				},
			},
		},
```

actualStr (has two principals)  = 
```
(*envoy_config_rbac_v2.Policy)(0xc0002a4000)(permissions:<and_rules:<rules:<or_rules:<rules:<header:<name:":method" exact_match:"GET" > > > > > > principals:<and_ids:<ids:<or_ids:<ids:<metadata:<filter:"istio_authn" path:<key:"source.principal" > value:<string_match:<exact:"td/ns/foo/sa/bar" > > > > ids:<metadata:<filter:"istio_authn" path:<key:"source.principal" > value:<string_match:<exact:"cluster.local/ns/foo/sa/bar" > > > > > > > > )
```

values (has one principal) = 
`[cluster.local/ns/foo/sa/bar]` 

This case should not pass, but it will without the `checkPrincipalCounts` function.